fixed gpg.decrypt function

leifliddy opened the pull request:
### What does this PR do?
 `gpg.decrypt` now works when `use_passphrase=True`

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/62806

### Previous Behavior
 `gpg.decrypt` did not function correctly when `use_passphrase=True` was passed

### New Behavior
 `gpg.decrypt` now functions correctly when `use_passphrase=True` was passed

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

leifliddy commented:
I'll fix the `gpg.encrypt` function (when `use_passphrase=True`) once this PR is approved https://github.com/saltstack/salt/pull/62896

MKLeb commented:
Hi @leifliddy, it seems like there was some discussion on the fix in https://github.com/saltstack/salt/pull/62896.  If this PR depends on that one, I will ask to get more eyes on the other one to see how we want to handle it.

leifliddy commented:
@MKLeb Cool, thanks! It seems to be a bit a divisive topic. If Gareth wants to "follow our normal deprecation path" I'd just like some clarification on what that means. There's no salt documentation that lists `gnupg` as a supported library so I'm sure how to deprecate something that's not supported. Could you please see what you can do?

MKLeb commented:
@leifliddy The GPG version check PR has been merged... were you planning on adding the encryption changes to this PR or a separate one?

leifliddy commented:
I can do it on this one. I just need a day or so to rebase the current changes and to add the encryption changes.

leifliddy committed:
reworked encrypt function

1. don't query pillar for passphrase when we're not signing a message
2. allow a text message to be encrypted, signed and written to disk
3. removed gpg_passphrase dict line

leifliddy committed:
added decrypt and encrypt tests

leifliddy committed:
add changelog

leifliddy committed:
fixed formatting errors

leifliddy committed:
fixed formatting errors v2

leifliddy committed:
fixed formatting errors v3

leifliddy commented:
I realize that this PR is all over the place. It's technically resolving an issue with the decrypt and encrypt functions. But, it also introduces the new argument `always_trust`
So without `always_trust`, `gpg.encrypt` will not encrypt a message with a key that's not ultimately trusted without prompting the user. So it's not possible to automate that without `always_trust=True` or by trusting the key via the `trust_key` function 
So I needed that `always_trust` in order for the unit test to complete. Plus, it's a useful argument to have anyway. 

I couldn't for the life of me develop a unit test that run through the `trust_key` function --which uses `cmd.run_all` to trust a key
```
res = __salt__["cmd.run_all"](cmd, stdin=stdin, runas=_user, python_shell=False)
```
I just didn't know how to mock that....

On another note, I don't see why we can't just use the `trust_keys` function from the `python-gnupg` module
https://github.com/vsajip/python-gnupg/blob/master/gnupg.py

That would allow the function to process multiple keys at once while not relying on `cmd.run_all` to do that. 
I'll save that for a future update though...

Anyways, let me know what you think...

leifliddy committed:
remove output check

If output=None, then key will be outputted to stdout

leifliddy commented:
Damn....I see four people had already approved this but I needed to make one last change.

```python
    if output:
        result = gpg.encrypt_file(
            _fp,
            recipients,
            passphrase=gpg_passphrase,
            sign=sign,
            always_trust=always_trust,
            output=output,
        )
```
But, there's no need to do a conditional check on `output` here.

Ch3LL commented:
No worries. We will wait to re-review after you push your changes

leifliddy commented:
Thanks. The change has already been pushed.

Ch3LL commented:
Oh I see, The commit was before your comment. Thanks, i'll review right now

Ch3LL commented on salt/modules/gpg.py:
```diff
@@ -1186,8 +1188,11 @@ def encrypt(
+        Passphrase is received from Pillar.
+
+    always_trust
+        Skip key validation and assume that used keys are fully trusted.
```
Can you add `.. versionadded:: 3006.0` to this new argument

Ch3LL reviewed (changes requested):
Just a minor request

leifliddy committed:
added versionadded line
